src/rgw/rgw_op.cc

```diff
@@ -3982,55 +3982,67 @@ void RGWPutMetadataBucket::execute()
-    cors_config.encode(bl);
-    emplace_attr(RGW_ATTR_CORS, std::move(bl));
-  }
+    if (has_cors) {
+      buffer::list bl;
+      cors_config.encode(bl);
+      emplace_attr(RGW_ATTR_CORS, std::move(bl));
+    }
 
-  /* It's supposed that following functions WILL NOT change any special
-   * attributes (like RGW_ATTR_ACL) if they are already present in attrs. */
-  prepare_add_del_attrs(s->bucket_attrs, rmattr_names, attrs);
-  populate_with_generic_attrs(s, attrs);
+    /* It's supposed that following functions WILL NOT change any special
+     * attributes (like RGW_ATTR_ACL) if they are already present in attrs. */
+    prepare_add_del_attrs(s->bucket_attrs, rmattr_names, attrs);
+    populate_with_generic_attrs(s, attrs);
 
-  /* According  to the Swift's behaviour and its container_quota WSGI middleware
-   * implementation: anyone with write permissions is able to set the bucket
-   * quota. This stays in contrast to account quotas that can be set only by
-   * clients holding reseller admin privileges. */
-  op_ret = filter_out_quota_info(attrs, rmattr_names, s->bucket_info.quota);
-  if (op_ret < 0) {
-    return;
-  }
+    /* According  to the Swift's behaviour and its container_quota WSGI middleware
+     * implementation: anyone with write permissions is able to set the bucket
+     * quota. This stays in contrast to account quotas that can be set only by
+     * clients holding reseller admin privileges. */
+    op_ret = filter_out_quota_info(attrs, rmattr_names, s->bucket_info.quota);
+    if (op_ret < 0) {
+      return;
+    }
 
-  if (swift_ver_location) {
-    s->bucket_info.swift_ver_location = *swift_ver_location;
-    s->bucket_info.swift_versioning = (! swift_ver_location->empty());
-  }
+    if (swift_ver_location) {
+      s->bucket_info.swift_ver_location = *swift_ver_location;
+      s->bucket_info.swift_versioning = (! swift_ver_location->empty());
+    }
 
-  /* Web site of Swift API. */
-  filter_out_website(attrs, rmattr_names, s->bucket_info.website_conf);
-  s->bucket_info.has_website = !s->bucket_info.website_conf.is_empty();
+    /* Web site of Swift API. */
+    filter_out_website(attrs, rmattr_names, s->bucket_info.website_conf);
+    s->bucket_info.has_website = !s->bucket_info.website_conf.is_empty();
 
-  /* Setting attributes also stores the provided bucket info. Due to this
-   * fact, the new quota settings can be serialized with the same call. */
-  op_ret = rgw_bucket_set_attrs(store, s->bucket_info, attrs,
-				&s->bucket_info.objv_tracker);
+    /* Setting attributes also stores the provided bucket info. Due to this
+     * fact, the new quota settings can be serialized with the same call. */
+    op_ret = rgw_bucket_set_attrs(store, s->bucket_info, attrs,
+				  &s->bucket_info.objv_tracker);
+
+    if (op_ret != -ECANCELED) {
+      break;
+    }
+
+    if (!store->try_refresh_bucket_info(s->bucket_info, nullptr,
+					&s->bucket_attrs)) {
+      break;
+    }
+  } while (--tries_remaining);
```

Comment:
so, the good news is that it seems like `RGWPutMetadataBucket::execute()` doesn't have any local variables. This means that instead of putting this whole block of code in a `do { } while`, you can put it in a separate method, and keep the `do { } while` here.